File: Packages/OmniaInfrastructure/Sources/OmniaInfrastructure/OpenAICompatibleClient.swift
import Foundation
import OmniaDomain

/// The OpenAI-compatible client over the transport seam (DES-010 §3.5).
///
/// The client constructs chat-completions requests for OpenAI-compatible
/// endpoints, decodes responses, and delivers streaming content, translating
/// every failure into `ProviderTransportError` (ARC-004, DES-009 §3.9). It is
/// internal to the package; the provider adapters consume it.
///
/// Authentication is by reference: the client resolves the credential through
/// the credential storage and builds the authorization header inside the
/// credential's scoped access, so the secret never enters logs, analytics, or
/// request metadata beyond the authorization header itself (ARC-001, ARC-005).
internal struct OpenAICompatibleClient: Sendable {
    private let transport: any ProviderTransport
    private let credentialStorage: any CredentialStorageProtocol

    init(
        transport: any ProviderTransport,
        credentialStorage: any CredentialStorageProtocol
    ) {
        self.transport = transport
        self.credentialStorage = credentialStorage
    }

    /// Performs a non-streaming chat-completions request and returns the
    /// decoded response.
    ///
    /// - Throws: `CredentialStorageError` when the credential cannot be
    ///   resolved, and `ProviderTransportError` for transport and decoding
    ///   failures.
    func chatCompletions(
        request: ChatCompletionRequest,
        endpoint: URL,
        credential: CredentialReference
    ) async throws -> ChatCompletionResponse {
        let httpRequest = try await makeRequest(
            request: request,
            endpoint: endpoint,
            credential: credential,
            streaming: false
        )
        let response = try await transport.send(httpRequest)
        do {
            return try Self.jsonDecoder.decode(ChatCompletionResponse.self, from: response.body)
        } catch {
            throw ProviderTransportError.invalidResponse
        }
    }

    /// Performs a streaming chat-completions request and returns the decoded
    /// chunk stream.
    ///
    /// The stream yields one `ChatCompletionChunk` per SSE data event and ends
    /// at the `[DONE]` marker or when the transport stream ends. Throws
    /// `CredentialStorageError` when the credential cannot be resolved.
    func streamChatCompletions(
        request: ChatCompletionRequest,
        endpoint: URL,
        credential: CredentialReference
    ) async throws -> AsyncThrowingStream<ChatCompletionChunk, any Error> {
        let httpRequest = try await makeRequest(
            request: request,
            endpoint: endpoint,
            credential: credential,
            streaming: true
        )
        return Self.chunkStream(from: transport.stream(httpRequest))
    }

    private func makeRequest(
        request: ChatCompletionRequest,
        endpoint: URL,
        credential: CredentialReference,
        streaming: Bool
    ) async throws -> ProviderHTTPRequest {
        let stored = try await credentialStorage.credential(for: credential)
        return try stored.withValue { apiKey in
            var request = request
            request.stream = streaming
            let body: Data
            do {
                body = try Self.jsonEncoder.encode(request)
            } catch {
                throw ProviderTransportError.invalidRequest
            }
            return ProviderHTTPRequest(
                url: endpoint.appendingPathComponent("chat/completions"),
                method: "POST",
                headers: [
                    "Content-Type": "application/json",
                    "Authorization": "Bearer \(apiKey)",
                ],
                body: body
            )
        }
    }

    private static func chunkStream(
        from chunks: AsyncThrowingStream<Data, any Error>
    ) -> AsyncThrowingStream<ChatCompletionChunk, any Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                var decoder = SSEDecoder()
                do {
                    for try await chunk in chunks {
                        if !emit(decoder.append(chunk), to: continuation) {
                            return
                        }
                    }
                    if !emit(decoder.finish(), to: continuation) {
                        return
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    /// Yields the decoded chunks of `events`; returns `false` after the `[DONE]`
    /// marker ends the stream.
    private static func emit(
        _ events: [SSEEvent],
        to continuation: AsyncThrowingStream<ChatCompletionChunk, any Error>.Continuation
    ) -> Bool {
        for event in events {
            if event.data.trimmingCharacters(in: .whitespacesAndNewlines) == "[DONE]" {
                continuation.finish()
                return false
            }
            do {
                continuation.yield(
                    try Self.jsonDecoder.decode(
                        ChatCompletionChunk.self,
                        from: Data(event.data.utf8)
                    )
                )
            } catch {
                continuation.finish(throwing: ProviderTransportError.invalidResponse)
                return false
            }
        }
        return true
    }

    private static var jsonEncoder: JSONEncoder {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        return encoder
    }

    private static var jsonDecoder: JSONDecoder {
        JSONDecoder()
    }
}

File: Android/core/network/src/main/kotlin/com/omnia/network/openai/OpenAICompatibleClient.kt
package com.omnia.network.openai

import com.omnia.domain.CredentialReference
import com.omnia.domain.CredentialStorageProtocol
import com.omnia.domain.ModelReference
import com.omnia.network.sse.SSEDecoder
import com.omnia.network.transport.ProviderHTTPRequest
import com.omnia.network.transport.ProviderTransport
import com.omnia.network.transport.ProviderTransportError
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.serialization.json.Json

/**
 * OpenAI-compatible client over the transport seam.
 *
 * Constructs chat-completions requests for OpenAI-compatible endpoints,
 * decodes responses, and delivers streaming content. Translates every
 * failure into [ProviderTransportError].
 *
 * Authentication is by reference: the client resolves the credential
 * through credential storage and builds the Authorization header inside
 * the credential's scoped access, so the secret never enters logs or
 * request metadata beyond the Authorization header.
 */
class OpenAICompatibleClient(
    private val transport: ProviderTransport,
    private val credentialStorage: CredentialStorageProtocol,
) {
    private val json = Json { ignoreUnknownKeys = true }

    /**
     * Non-streaming chat completions.
     */
    suspend fun chatCompletions(
        request: OpenAIChatCompletionRequest,
        endpoint: String,
        credential: CredentialReference,
    ): OpenAIChatCompletionResponse {
        val httpRequest = makeRequest(request, endpoint, credential, streaming = false)
        val response = transport.send(httpRequest)
        return try {
            json.decodeFromString<OpenAIChatCompletionResponse>(String(response.body))
        } catch (_: Exception) {
            throw ProviderTransportError.invalidResponse
        }
    }

    /**
     * Streaming chat completions. Returns a Flow of decoded chunks.
     */
    fun streamChatCompletions(
        request: OpenAIChatCompletionRequest,
        endpoint: String,
        credential: CredentialReference,
    ): Flow<OpenAIChatCompletionChunk> = flow {
        val httpRequest = makeRequest(request, endpoint, credential, streaming = true)
        val decoder = SSEDecoder()

        transport.stream(httpRequest).collect { chunk ->
            val events = decoder.append(chunk)
            for (event in events) {
                val trimmed = event.data.trim()
                if (trimmed == "[DONE]") return@collect
                try {
                    val decoded = json.decodeFromString<OpenAIChatCompletionChunk>(event.data)
                    emit(decoded)
                } catch (_: Exception) {
                    throw ProviderTransportError.invalidResponse
                }
            }
        }

        for (event in decoder.finish()) {
            val trimmed = event.data.trim()
            if (trimmed == "[DONE]") break
            try {
                val decoded = json.decodeFromString<OpenAIChatCompletionChunk>(event.data)
                emit(decoded)
            } catch (_: Exception) {
                throw ProviderTransportError.invalidResponse
            }
        }
    }.flowOn(Dispatchers.IO)

    /**
     * GET /models — returns model ID list.
     * Model-list records prove identity only; no inferred capabilities.
     */
    suspend fun models(
        endpoint: String,
        credential: CredentialReference,
    ): List<String> {
        val stored = credentialStorage.credential(credential)
        val httpRequest = stored.withValue { apiKey ->
            ProviderHTTPRequest(
                url = "${endpoint.trimEnd('/')}/models",
                method = "GET",
                headers = mapOf("Authorization" to "Bearer $apiKey"),
            )
        }
        val response = transport.send(httpRequest)
        val decoded = try {
            json.decodeFromString<OpenAIModelListResponse>(String(response.body))
        } catch (_: Exception) {
            throw ProviderTransportError.invalidResponse
        }
        return OpenAIMapping.modelIds(decoded)
    }

    /**
     * Probes whether the endpoint is reachable and authenticated.
     */
    suspend fun probeAvailability(
        endpoint: String,
        credential: CredentialReference,
    ): Boolean {
        return try {
            models(endpoint, credential)
            true
        } catch (_: Exception) {
            false
        }
    }

    private suspend fun makeRequest(
        request: OpenAIChatCompletionRequest,
        endpoint: String,
        credential: CredentialReference,
        streaming: Boolean,
    ): ProviderHTTPRequest {
        val stored = credentialStorage.credential(credential)
        return stored.withValue { apiKey ->
            val body = try {
                json.encodeToString(
                    OpenAIChatCompletionRequest.serializer(),
                    request.copy(stream = streaming)
                ).toByteArray(Charsets.UTF_8)
            } catch (_: Exception) {
                throw ProviderTransportError.invalidRequest
            }
            ProviderHTTPRequest(
                url = "${endpoint.trimEnd('/')}/chat/completions",
                method = "POST",
                headers = mapOf(
                    "Content-Type" to "application/json",
                    "Authorization" to "Bearer $apiKey",
                ),
                body = body,
            )
        }
    }
}
